packages/winnow/src/sql-query-builder/where-builder/index.js

```diff
@@ -0,0 +1,187 @@
+const { InvalidWhereParameterError } = require('../../errors');
+const convertSqlWhereToJsonWhere = require('./to-json-where');
+
+const timestampCastRegex = /(TIMESTAMP|DATE) '([^']*)'/gi; // eslint-disable-line
+const timestampWithoutZoneRegex =
+  /^\d{4}-\d{2}-\d{2} {0,1}\d{0,2}:{0,1}\d{0,2}:{0,1}\d{0,2}$/;
+
+// RegExp for name-first predicate, e.g "properties->`OBJECTID` = 1234"
+const fieldFirstObjectIdPredicateRegex =
+  /(properties|attributes)->`OBJECTID` (=|<|>|<=|>=) ([0-9]+)/g;
+
+// RegExp for value-first predicate, e.g "1234 = properties->`OBJECTID`""
+const valueFirstObjectIdPredicateRegex =
+  /([0-9]+) (=|<|>|<=|>=) (properties|attributes)->`OBJECTID`/g;
+
+const operatorInversions = {
+  '=': '=',
+  '>=': '<',
+  '>': '<=',
+  '<=': '>',
+  '<': '>=',
+};
+
+const SINGLE_QUOTE_ESCAPE = '\\';
+
+class WhereBuilder {
+  static create (params) {
+    const builder = new WhereBuilder(params);
+    return builder
+      .castTimestamps()
+      .replaceSingleQuoteEscapeChars()
+      .toJsonSQL()
+      .replaceObjectIdPredicates()
+      .addGeometryPredicate()
+      .value();
+  }
+
+  #geometryFilter;
+  #options;
+  #where;
+  #geometryPredicate;
+
+  constructor ({ where, geometry, ...options } = {}) {
+    this.#where = where;
+    this.#geometryFilter = geometry;
+    this.#geometryPredicate;
+    this.#options = options;
+  }
+
+  addGeometryPredicate () {
+    if (!this.#geometryFilter) {
+      return this;
+    }
+
+    const spatialPredicate = this.#options.spatialPredicate || 'ST_Intersects';
+    
+    // The "?" in the string below is a SQL query parameter.  When it is executed,
+    // a supplied value is used in its place
+    this.#geometryPredicate = `${spatialPredicate}(geometry, ?)`;
+
+    return this;
+  }
+
+  #shouldReplaceObjectIdPredicates() {
+    return this.#where?.includes('OBJECTID') && !this.#options?.idField;
+  }
+
+  castTimestamps () {
+    if (!this.#where) {
+      return this;
+    }
+
+    this.#where = this.#where.replace(timestampCastRegex, function(match, cast, timestamp){
+      return `'${normalizeTimestamp(timestamp)}'`;
+    });
+
+    return this;
+  }
+
+  replaceSingleQuoteEscapeChars() {
+    if (!this.#where) {
+      return this;
+    }
+
+    let insideSingleQuote = false;
+    
+    const charArr = this.#where.split('');
+  
+    charArr.forEach((currentChar, i) => {
+      const nextChar = charArr[i + 1];
+      const prevChar = charArr[i - 1];
+      const lastTwoChars = `${charArr[i - 1]}${charArr[i - 0]}`;
+  
+      if (isBeginningOfSingleQuoteWrap(currentChar, insideSingleQuote)) {
```

Comment:
Nit: This function might benefit from a few simple comments. I think the output/functionality/meaning of some of these conditionals, which have been decomposed to their own methods, aren't entirely clear when just reading the method name, so I find myself jumping around this file quite a bit to parse this all. Maybe just a top level comment of a before or after string thats undergone replaceSingleQuoteEscapeChars.

Reply:
yeah, that's all fair.  This function replaces SQL single quote escape sequences; so:

```sql
WHERE foo = 'something with a single quote like ''this'', you know?'
```

Unfortunately, most SQL uses single quotes as the escape character for single quotes (lol, why did they do that?).  But oddly, our SQL parser doesn't understand that and is instead expecting them to be escaped by a `\`. So:

```sql
WHERE foo = 'something with a single quote like \'this\', you know?'
```

This method is essentially a state machine that reads the SQL string determines if we are already in the outer single quote and then identifies any double single quote sequences (`''`) and replaces them with `\'`.  Went with a state machine because the regex was crazy and hard to handle edge cases.

Will add comments.